presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java

```diff
@@ -0,0 +1,513 @@
+    @SqlType(GEOMETRY)
+    public static Slice stPolygon(@SqlType(StandardTypes.VARCHAR) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(OGCGeometry.fromText(input.toStringUtf8()), "st_polygon", POLYGON);
+        return serialize(geometry);
+    }
+
+    @Description("Returns the area of the geometry if it is a polygon or multi-polygon.")
+    @ScalarFunction("st_area")
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stArea(@SqlType(GEOMETRY) Slice input)
+    {
+        return deserialize(input).getEsriGeometry().calculateArea2D();
+    }
+
+    @Description("Returns Geometry of input WKT text")
+    @ScalarFunction("st_geometry_from_text")
+    @SqlType(GEOMETRY)
+    public static Slice stGeometryFromText(@SqlType(StandardTypes.VARCHAR) Slice input)
+    {
+        return serialize(OGCGeometry.fromText(input.toStringUtf8()));
+    }
+
+    @SqlNullable
+    @Description("Returns wtk text of input Geometry")
+    @ScalarFunction("st_as_text")
+    @SqlType(StandardTypes.VARCHAR)
+    public static Slice stAsText(@SqlType(GEOMETRY) Slice input)
+    {
+        return Slices.utf8Slice(deserialize(input).asText());
+    }
+
+    @Description("Returns point that is the center of the polygon's envelope")
+    @ScalarFunction
+    @SqlType(GEOMETRY)
+    public static Slice stCentroid(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "st_centroid", POLYGON);
+        SpatialReference reference = geometry.getEsriSpatialReference();
+        Envelope envelope = getEnvelope(geometry);
+        Point centroid = new Point((envelope.getXMin() + envelope.getXMax()) / 2, (envelope.getYMin() + envelope.getYMax()) / 2);
+        return serialize(OGCGeometry.createFromEsriGeometry(centroid, reference));
+    }
+
+    @Description("Returns count of coordinate components")
+    @ScalarFunction
+    @SqlType(StandardTypes.BIGINT)
+    public static long stCoordinateDimension(@SqlType(GEOMETRY) Slice input)
+    {
+        return deserialize(input).coordinateDimension();
+    }
+
+    @Description("Returns spatial dimension of geometry")
+    @ScalarFunction
+    @SqlType(StandardTypes.BIGINT)
+    public static long stDimension(@SqlType(GEOMETRY) Slice input)
+    {
+        return deserialize(input).dimension();
+    }
+
+    @SqlNullable
+    @Description("Returns true if and only if the line is closed")
+    @ScalarFunction("st_is_closed")
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean stIsClosed(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "st_is_closed", LINE, MULTILINE);
+        MultiPath lines = (MultiPath) geometry.getEsriGeometry();
+        int pathCount = lines.getPathCount();
+        boolean result = true;
+        for (int i = 0; result && i < pathCount; i++) {
+            Point start = lines.getPoint(lines.getPathStart(i));
+            Point end = lines.getPoint(lines.getPathEnd(i) - 1);
```

Comment:
I don't know much about the getPoint/getPathEnd APIs, but shouldn't this be `getPathEnd(i - 1)` ?

Reply:
getPathEnd: Returns the index immediately following the last index of the path. So use MultiPath.getPathEnd(i) -1 to get the last index of the path